baybe/campaign.py

```diff
@@ -82,6 +82,21 @@ class Campaign(SerialMixin):
     numerical_measurements_must_be_within_tolerance: bool = field(default=None)
     """Deprecated! Raises an error when used."""
 
+    def __str__(self) -> str:
+        start_bold = "\033[1m"
+        end_bold = "\033[0m"
+
+        # Get str representation of campaign objects
+        fields_to_print = [self.searchspace, self.objective, self.recommender]
+        fields_str = "\n".join(str(x) for x in fields_to_print)
+
+        # Put all relevant attributes of the campaign in one string
+        campaign_str = f"""\n{start_bold}_____ Campaign _____
```

Comment:
@Scienfitz Im not sure which other symbols you mean. Can you please give me any suggestions? I will try them and send you the screenshots.

Reply:
currently using all kinds of stuff ie `___` or `***` or `|-->`
![image](https://github.com/emdgroup/baybe/assets/17951239/e1984121-2cba-48d8-8e31-446e6f07d386)

Reply:
Aah i understand what you mean. I used `*` for the headers and `|-->` for subpoins, but did not know what to use for campaign as it is another header above `searchspace`. But maybe I should just use `*` again to make all headers look the same.

Reply:
yeah we cant really make them ordered like that, sicne you could also only print a searchspace on its own, theres no sense of the hirarchy

Why not use symbols like `###` with increasing amount like in md file;s?

ie
```
# Campaign
...
## Search Space
...
```

you could also pre-pend the `#` in front of all lines of the lower class str representaitons so it would be passed through, but i dont know whether that would look good

Reply:
This makes sense. It will look like the changelog file then. But this requires modifying the `str` of the discrete and continuous search space as well to remove the already existing symbols. I will do it, if its okay for this PR

Reply:
Would be fine for me :)

Reply:
Updated output: 
![image](https://github.com/emdgroup/baybe/assets/118910837/09f10cea-e4b5-4329-bfdc-47dfbc58eb19)
![image](https://github.com/emdgroup/baybe/assets/118910837/2b92ff0b-8ce4-461e-b583-3dc6bb810542)

Reply:
That looks good :) For me, that would already work. We'd then have that a `search space` would start on the second level when being printed on its own but that is something that doesn't bother me personally since the relative hierarchy is still clear.

Reply:
Looking at the branch, it also seems like this is already properly rebased. @Scienfitz if you are happy with this solution feel free to resolve and merge.

Reply:
@AVHopp Yes exactly. This was my only concern, but it would be very complex to solve in another way. As you said the hierarchy is still clear and it is clear that there is another layer above `search space`.

Reply:
i dont see why it would be complciated to have the hirarchical sue of `#` dynamical and not hatdcoded.

1. In each objectss print function you print custom info (which you will summarize with a simple `#` in front of it)
2. Additionally you can the `__str__` of subordinate things, since these will also have their own `#`s implemented, all you need to do is to add one `#` in front of the text you get from that `__str__`. 
3. If properly implemented in all str fnctions each time you print any object it wil automatically have the right amounts of `#`s to correctly indicate the hirarchy
4. One could also try to add the `#` in front of all lines of an incoming `__str__` instead of just the very top one, but one would have to check how that looks

Reply:
@Scienfitz The solution you suggested sounds very good. But unfortuonately it does not deliver the dynamic output we need, because:
- Imagine you are printing a campaign object. You will add a `#` for the campaign itself and another `#` for every `__str__` you call (which is the one for `search space`, `objective` and `recommender`. 
![image](https://github.com/emdgroup/baybe/assets/118910837/816cc0bf-c948-4376-bc1d-8d3adaf9c8f3)
 the `__str__` of the `search space` has its own  `#` and an additional one for `discrete` and `continuous`:
![image](https://github.com/emdgroup/baybe/assets/118910837/64ea7c00-b93f-48cd-a10b-82219b27a24a)
But now we have two `#` for the `search space` and the `discrete space` each, and this solution does not indicate whether the `__str__` of `search space` was called insid campaign or individually and will deliver this output:
![image](https://github.com/emdgroup/baybe/assets/118910837/87089d8a-4af7-4626-a346-da5f5239eaa3)

- Plus: in this solution you will get a `#` infront of an empty space like this: 
![image](https://github.com/emdgroup/baybe/assets/118910837/3ef9e4df-54d4-4317-9a0a-3192c914b379)
 (the `continuous` space is empty)

(please tell me if I understood you correctly or implemented something else :)

Reply:
and what about adding it in front of every line of an incoming `__str__` instead of just the beginning

Reply:
what do mean exactly by every incoming line? In campaign I already put `#` in front of every incoming `__str__` but they are only 3, because it calls the `__str__` of the sub objects internally.

Reply:
But how will this help the dynamic printing? Currently I am printing the `#` only for the headers, but if we want to print it infront of every line the output will look like this: 
![image](https://github.com/emdgroup/baybe/assets/118910837/82a9b30f-8543-43b0-a1b1-c9b89447d0fd)

or do mean to completly remove the hierarchy and simply unify all lines with one `#`? 
![image](https://github.com/emdgroup/baybe/assets/118910837/c50f8c45-a19d-464c-8d5f-421177e458d7)

@AVHopp what do you think?

Reply:
I do not think that having a "#" in front of *every* line looks good. I'd actually prefer a hierarchical version with more and more "#" the deeper you go down in hierarchy.
I also think that this can be achieved rather easily: The `__str__()` method of each object contains "First level logic", i.e., it starts with a single "#" for the main head, "##" for the next and so on. Then, when calling this function in the next-higher level (e.g. calling `__str__()` of a search space within campaign), we just do `str(self.searchspace).replace("# ", "## ")` to replace the final # in each line by two.

Reply:
Alex repeated exactly my idea, though i think your implementation suggestion would double the amount of # in each level and there might be # in the strings that you dont watnt to extend (ie in var names)

In the pictures you still havent added it in front of every line. Keep in mind that ilI also suggested it could be a tab or spaces put in front of every line to create an indentation

try '.replace("\n", "\n  ")' or so

Reply:
I did not repeat your idea since my vote is to NOT have it in every line but only in headers. I honestly do not see the reason of putting one # sign in absolutely every line, why do you want to have this?

Reply:
I used the `replace` function and it turned out pretty good. The hierarchy that we had in the begining is now dynamic and  I added an empty check, so that no `#` will be printed for empty spaces.
`print(campaign)`:
![image](https://github.com/emdgroup/baybe/assets/118910837/fe74e713-2214-40fb-aeb5-1891805a15eb)
`print(campaign.searchspace)`:
![image](https://github.com/emdgroup/baybe/assets/118910837/00431ad9-ed7f-437b-a820-f0ce7f0bec39)

Is this a good solution for both of you?

Reply:
this is the best look so far but as I wrote it will double the amount of #'s and also double them where its not intended eg in all texts

would also like to look at the indented version, we can throw it out if it looks bad

Reply:
For the intended version, you could probably use a `repalce("\n", "\n\t")` right? That would also avoid the issue with the multiple # or accidentialy replacing stuff that shouldn't be replaced, right?

Reply:
This would be the result of replacing the `\n, \n\t`:
![image](https://github.com/emdgroup/baybe/assets/118910837/f8f2de03-f31a-42b6-a1ab-8fda98c72700)

But I think that a `\t` was too much space, so tried a simple one space instead `\n `:
![image](https://github.com/emdgroup/baybe/assets/118910837/f40b3ef4-3c20-4916-8c65-eae7dc054082)

Reply:
space looks `gucci`
lets also see a screenshot with 2 spaces, but Id probably choose the space version with 1 or 2 spaces per level
thats fine for me to move on and merge the PR

Reply:
missing space: 
![image](https://github.com/emdgroup/baybe/assets/17951239/cf152dce-d06e-48ac-8923-a39d0c899e09)

Reply:
if alex is also fine with the space ill leave all choices for him and ill greenlight the PR

Reply:
This is the 2 spaces version: 
![image](https://github.com/emdgroup/baybe/assets/118910837/992cad43-1958-49c8-a232-219f86031b65)

Reply:
maybe lets go with 1 space to save overall space but id be fine with both

Reply:
additional newline thats not present in other components: 
![image](https://github.com/emdgroup/baybe/assets/17951239/3b1e5ffe-2850-4e61-b895-96adf4b9c41a)

Reply:
@AVHopp what do you think?